faunadb-java-dsl/src/main/java/com/faunadb/client/types/Constructors.java

```diff
@@ -0,0 +1,236 @@
+      this.parameterNames = getParameterNames(constructor.getParameterAnnotations());
+      this.parameterTypes = getParameterTypes(constructor.getGenericParameterTypes());
+      this.writeProperties = filterProperties(Properties.getWriteProperties(constructor.getDeclaringClass()), this.parameterNames);
+    }
+
+    AbstractConstructor(Method method) {
+      this.rawClass = method.getDeclaringClass();
+      this.parameterNames = getParameterNames(method.getParameterAnnotations());
+      this.parameterTypes = getParameterTypes(method.getGenericParameterTypes());
+      this.writeProperties = filterProperties(Properties.getWriteProperties(method.getDeclaringClass()), this.parameterNames);
+    }
+
+    private SimpleType[] getParameterTypes(Type[] genericParameterTypes) {
+      SimpleType[] parameterTypes = new SimpleType[genericParameterTypes.length];
+
+      for (int i = 0; i < genericParameterTypes.length; i++) {
+        parameterTypes[i] = Types.of(genericParameterTypes[i]);
+      }
+
+      return parameterTypes;
+    }
+
+    private static Property[] filterProperties(Property[] properties, String[] names) {
+      if (names.length == 0)
+        return properties;
+
+      final Set<String> namesToRemove = newHashSet(names);
+      return filter(newArrayList(properties), new Predicate<Property>() {
+        @Override
+        public boolean apply(Property input) {
+          return !namesToRemove.contains(input.getName());
+        }
+      }).toArray(new Property[0]);
+    }
+
+    private String[] getParameterNames(Annotation[][] parameterAnnotations) {
+      String[] parameterNames = new String[parameterAnnotations.length];
+
+      for (int i = 0; i < parameterAnnotations.length; i++) {
+        Annotation[] annotations = parameterAnnotations[i];
+
+        for (Annotation annotation : annotations) {
+          if (annotation.annotationType() == FaunaField.class) {
+            parameterNames[i] = ((FaunaField) annotation).value();
+          }
+        }
+
+        if (parameterNames[i] == null)
+          throw new FaunaException(format("All constructor or factory method arguments must be annotated with @%s", FaunaField.class.getSimpleName()));
+      }
+
+      return parameterNames;
+    }
+
+    @Override
+    public Result<Object> decode(Value value) {
+      try {
+        ObjectV object = (ObjectV) value;
+        Object instance = newInstance(buildArguments(object));
+
+        for (Property property : writeProperties) {
+          Object propertyValue = decodeImpl(object.values.get(property.getName()), property.getType()).get();
+          property.set(instance, propertyValue);
+        }
+
+        return Result.success(instance);
+      } catch (Exception ex) {
+        return Result.fail(format("Could not instantiate object of class %s", rawClass.getName()), ex);
+      }
+    }
+
+    @Override
+    public Result<Value> encode(Object value) {
+      return Result.fail("Not supported");
```

Comment:
Well, there are some constraints that is valid for one that is not valid for the other, for example, if you have a class meant to be used only for encoding that class would have only attributes annotated but might not have a constructor configured, that answer the above question as well, this is why I keep two separate maps of codecs.

If we decide to force every class to be either encodable and decodable than we keep only one map and implement a single codec for objects